sls-versions/src/main/java/com/palantir/sls/versions/MatchResult.java

```diff
@@ -26,15 +26,28 @@
     int groupCount();
 
     final class RegexMatchResult implements MatchResult {
+        private static final int RADIX = 10;
+        private final String string;
         private final Matcher matcher;
 
-        RegexMatchResult(Matcher matcher) {
+        RegexMatchResult(String string, Matcher matcher) {
+            this.string = string;
             this.matcher = matcher;
         }
 
         @Override
         public int groupAsInt(int group) {
-            return Integer.parseInt(matcher.group(group));
+            int groupStart = matcher.start(group);
+            int groupEnd = matcher.end(group);
+            if (groupStart == -1) {
+                throw new NumberFormatException();
+            }
+
+            if (groupEnd - groupStart == 1) {
+                return Character.digit(string.codePointAt(groupStart), RADIX);
+            }
+
+            return Integer.parseUnsignedInt(string, groupStart, groupEnd, RADIX);
```

Comment:
Doe the first condition actually perform better than `Integer.parseUnsignedInt`? You'd think that sort of optimization would be built in. Can't crack open the implementation on thumbs though.

Reply:
fair point that it's probably ok to ditch the single return Character.digit thing... it only avoids a few if statements which are probably trivially branch predicted anyway

```
public static int parseInt(CharSequence s, int beginIndex, int endIndex, int radix)
                throws NumberFormatException {
        Objects.requireNonNull(s);

        if (beginIndex < 0 || beginIndex > endIndex || endIndex > s.length()) {
            throw new IndexOutOfBoundsException();
        }
        if (radix < Character.MIN_RADIX) {
            throw new NumberFormatException("radix " + radix +
                                            " less than Character.MIN_RADIX");
        }
        if (radix > Character.MAX_RADIX) {
            throw new NumberFormatException("radix " + radix +
                                            " greater than Character.MAX_RADIX");
        }

        boolean negative = false;
        int i = beginIndex;
        int limit = -Integer.MAX_VALUE;

        if (i < endIndex) {
            char firstChar = s.charAt(i);
            if (firstChar < '0') { // Possible leading "+" or "-"
                if (firstChar == '-') {
                    negative = true;
                    limit = Integer.MIN_VALUE;
                } else if (firstChar != '+') {
                    throw NumberFormatException.forCharSequence(s, beginIndex,
                            endIndex, i);
                }
                i++;
                if (i == endIndex) { // Cannot have lone "+" or "-"
                    throw NumberFormatException.forCharSequence(s, beginIndex,
                            endIndex, i);
                }
            }
            int multmin = limit / radix;
            int result = 0;
            while (i < endIndex) {
                // Accumulating negatively avoids surprises near MAX_VALUE
                int digit = Character.digit(s.charAt(i), radix);
                if (digit < 0 || result < multmin) {
                    throw NumberFormatException.forCharSequence(s, beginIndex,
                            endIndex, i);
                }
                result *= radix;
                if (result < limit + digit) {
                    throw NumberFormatException.forCharSequence(s, beginIndex,
                            endIndex, i);
                }
                i++;
                result -= digit;
            }
            return negative ? result : -result;
        } else {
            throw NumberFormatException.forInputString("", radix);
        }
    }
```